### PR TITLE
CVSL-2271 update hdcad on sentence change events

### PR DESCRIPTION
**Pull request by louisenorris**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceFunctions.kt [hunk 1]
```diff
@@ -26,6 +28,14 @@ fun Licence.getSentenceChanges(newSentence: UpdateSentenceDatesRequest): Sentenc
   val tusedChanged =
     nullableDatesDiffer(newSentence.topupSupervisionExpiryDate, this.topupSupervisionExpiryDate)
   val prrdChanged = nullableDatesDiffer(newSentence.postRecallReleaseDate, this.postRecallReleaseDate)
+
+  if (this is HdcLicence) {
+    val hdcadChanged = nullableDatesDiffer(newSentence.homeDetentionCurfewActualDate, this.homeDetentionCurfewActualDate)
+
+    val isMaterial = (lsdChanged || ledChanged || tussdChanged || tusedChanged || prrdChanged || hdcadChanged || (sedChanged && this.statusCode == LicenceStatus.APPROVED))
+    return SentenceChanges(lsdChanged, ledChanged, sedChanged, tussdChanged, tusedChanged, prrdChanged, isMaterial, hdcadChanged)
+  }
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceFunctions.kt:37]: **suggestion:** this is quite a lot to duplicate, you could do:
```
val hdcadChanged = this is HdcLicence && nullableDatesDiffer(newSentence.homeDetentionCurfewActualDate, this.homeDetentionCurfewActualDate)
```
and then always pass through to the `isMaterialCheck` and `SentenceChanges` (which would then mean you don't need to provide a default arg to it)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/UpdateSentenceDateService.kt [hunk 2]
```diff
@@ -170,6 +178,9 @@ class UpdateSentenceDateService(
         append("TUSSD ${sentenceDatesRequest.topupSupervisionStartDate} ")
         append("TUSED ${sentenceDatesRequest.topupSupervisionExpiryDate} ")
         append("PRRD ${sentenceDatesRequest.postRecallReleaseDate}")
+        if (licenceEntity is HdcLicence) {
+          append("HDCAD ${sentenceDatesRequest.homeDetentionCurfewActualDate} ")
+        }
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/UpdateSentenceDateService.kt:183]: based on our discussion, this is all cool and good we're limiting the amount of type-checking to this specific case  👍

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceFunctions.kt (hunk 1 above)
Comment (louisenorris) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceFunctions.kt]: Cool, refactored in the latest commit

**Review by andrewrlee — APPROVED** [review]
LGTM

**Pull request merged**